pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -98,7 +98,8 @@ public boolean add(MessageImpl<?> msg, SendCallback callback) {
                 messageMetadata.setSequenceId(msg.getSequenceId());
                 lowestSequenceId = Commands.initBatchMessageMetadata(messageMetadata, msg.getMessageBuilder());
                 this.firstCallback = callback;
-                batchedMessageMetadataAndPayload = allocator.compositeBuffer();
+                batchedMessageMetadataAndPayload = allocator.buffer(
+                        Math.min(maxBatchSize, ClientCnx.getMaxMessageSize()));
```

Comment:
How can you guarantee that this buffer won't overflow?

BTW, it's interesting that `maxBatchSize` actually unused before this change :)

Reply:
i.e., we need some explanation for the math on calculating buffer size.

Reply:
This changing is just revert the pr https://github.com/apache/pulsar/pull/15033.  PR#15033 is also try to solve the memory waste problem but bring a significant performance regression @tisonkun

Reply:
> How can you guarantee that this buffer won't overflow?

`batchedMessageMetadataAndPayload` will resize and do memory copy if the space is not enough